lametro/paired_event_stream.py

```diff
@@ -0,0 +1,225 @@
+import datetime
+import logging
+from typing import Generator
+
+from legistar.events import LegistarAPIEventScraper
+
+LOGGER = logging.getLogger(__name__)
+
+
+class LAMetroAPIEvent(dict):
+    """
+    This class is for adding methods to the API event dict
+    to faciliate maching events with their other-language
+    partners.
+    """
+
+    @property
+    def is_spanish(self):
+        return self["EventBodyName"].endswith(" (SAP)")
+
+    @property
+    def own_key(self):
+        return (self["EventBodyName"], self["EventDate"])
+
+    @property
+    def _partner_name(self):
+        if self.is_spanish:
+            return self["EventBodyName"].rstrip(" (SAP)")
+        else:
+            return self["EventBodyName"] + " (SAP)"
+
+    def is_partner(self, other):
+        return (
+            self._partner_name == other["EventBodyName"]
+            and self["EventDate"] == other["EventDate"]
+        )
+
+    @property
+    def partner_search_string(self):
+        search_string = "EventBodyName eq '{}'".format(self._partner_name)
+        search_string += " and EventDate eq datetime'{}'".format(self["EventDate"])
+
+        return search_string
+
+    @property
+    def partner_key(self):
+        return (self._partner_name, self["EventDate"])
+
+    @property
+    def key(self):
+        return (self["EventBodyName"], self["EventDate"])
+
+
+class LAMetroWebEvent(dict):
+    """
+    This class is for adding methods to the web event dict
+    to facilitate labeling and sourcing audio appropriately.
+    """
+
+    @property
+    def has_audio(self):
+        return self["Meeting video"] != "Not\xa0available"
+
+    @property
+    def has_ecomment(self):
+        return self["eComment"] != "Not\xa0available"
+
+
+class LAMetroAPIEventScraper(LegistarAPIEventScraper):
+    BASE_URL = "https://webapi.legistar.com/v1/metro"
+    WEB_URL = "https://metro.legistar.com/"
+    EVENTSPAGE = "https://metro.legistar.com/Calendar.aspx"
+    TIMEZONE = "America/Los_Angeles"
+
+
+class PairedEventStream:
+    """
+    - Input: Stream of events that need to be paired
+        - Deduplicate events
+        - Pair events
+            - From stream
+            - From API (partner search)
+        - Merge events
+    - Output: Stream of events that have been merged
+    """
+
+    def __init__(self, events: list[dict], find_missing_partner: bool = True) -> None:
+        self.events = [LAMetroAPIEvent(event) for event in events]
+        self.find_missing_partner = find_missing_partner
+
+    def __iter__(
```

Comment:
This is motivating me to implement more iterators in the future! Great use of type hints too.